### PR TITLE
Interface and converter

### PR DESCRIPTION
**Pull request by lnguyen2693**

### CODE REVIEW

File: src/app/types/db.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  onDuty: string[];
+  tags: Tags;
+}
+
+export interface Question {
+  title: string;
+  description: string;
+  public: boolean;
+  // should state be enum or boolean or string?
+  state: string;
```
Comment (nickbar01234) [inline src/app/types/db.ts]: I think it should be an enum here.

```ts
enum QuestionState {
  ...
}
```

File: src/app/utils/firestore.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      public: question.public,
+      state: question.state,
+      timestamp: question.timestamp,
+      group: question.group,
+      tags: question.tags,
+    };
+  },
+  fromFirestore(snapshot: QueryDocumentSnapshot, options: any): Question {
+    const data = snapshot.data(options);
+    if (!data) {
```
Comment (nickbar01234) [inline src/app/utils/firestore.ts]: When would there be a case where `data` is null? Maybe we can do

```ts
return { ... } as Question
```

File: src/app/types/db.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  color: string;
+  colorFill: boolean;
+}
+
+export interface Course {
+  name: string;
+  location: string;
+  professors: string[];
+  students: string[];
+  TAs: string[];
```
Comment (nickbar01234) [inline src/app/types/db.ts]: nit: `tas` instead of `TAs`

File: src/app/utils/firestore.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+export const questionConverter = {
+  toFirestore(question: Question): DocumentData {
+    return {
+      title: question.title,
+      description: question.description,
+      public: question.public,
+      state: question.state,
+      timestamp: question.timestamp,
```
Comment (nickbar01234) [inline src/app/utils/firestore.ts]: I think timestamp should be generated on server side. If we generate it on client, it's possible to modify the clock

Comment (nickbar01234) [reply src/app/utils/firestore.ts]: https://stackoverflow.com/questions/69519447/how-to-get-server-timestamp-from-firebase-v9

File: src/app/utils/firestore.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import { Course, Question } from "../types/db";
+import { DocumentData, QueryDocumentSnapshot } from "firebase/firestore";
+
+export const courseConverter = {
+  toFirestore(course: Course): DocumentData {
```
Comment (nickbar01234) [inline src/app/utils/firestore.ts]: I think having the type to be `Course` is a bit restricting for end-users, i.e they must always pass in the full-object. So `PartialWithFieldValue<Course>` might make more sense here. Similar comment for `Question`.

File: src/app/utils/firestore.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import { Course, Question } from "../types/db";
```
Comment (nickbar01234) [inline src/app/utils/firestore.ts]: Side note, I think the converter should be in a dedicated folder. Maybe `services/firestore`?

**Review by nickbar01234 — CHANGES_REQUESTED** [review]

File: src/app/utils/firestore.ts (hunk 2 above)
Comment (johnny-t06) [reply src/app/utils/firestore.ts]: +1 I like that idea. 
Looks like `data` might be null if 1) the document doesn't exist or 2) the user is not authenticated/have the right perms
https://stackoverflow.com/questions/53607050/documentsnapshot-is-always-null-for-a-valid-document-reference

**Review by johnny-t06 — CHANGES_REQUESTED** [review]
Looks great, I am just echoing Nick's comments

File: src/app/types/db.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  name: string;
+  location: string;
+  professors: string[];
+  students: string[];
+  tas: string[];
+  onDuty: string[];
+  tags: Tags;
+}
+enum QuestionState{
+  "pending" = 1,
```
Comment (nickbar01234) [inline src/app/types/db.ts:31]: nit: Should be capitalized

**Pull request merged**